### PR TITLE
8232722: G1 archive region deallocation may shrink the heap below -Xms

### PR DESCRIPTION
**Pull request by tschatzl**
Hi all,

  can I have reviews for this change that properly deallocates/uncommits CDS archive regions when failing to load the CDS archive?

In particular this caused the nuisance mentioned in the CR where even if -Xms==-Xmx, g1 uncommitted the heap memory anyway.

Testing: gha, manual testing as below

There is no (existing) way to induce CDS load errors easily, so what I did was adding `-XX:+UseNewCode` in `filemap.cpp:2202` to simulate failures when enabled. Obviously I removed the flag in this change.

Here's the problematic case:

```
$java -XX:GCCardSizeInBytes=128 -Xmx128m -Xms128m -Xlog:gc+region=trace,gc+ergo+heap=debug -XX:+UseNewCode -version

[0.050s][trace][gc,region   ] G1HR ALLOC(OLD) [0x00000000ffe00000, 0x00000000fff00000, 0x00000000fff00000]
[0.050s][trace][gc,region   ] G1HR ALLOC(OLD) [0x00000000fff00000, 0x00000000fff06278, 0x0000000100000000]
[0.050s][trace][gc,region   ] G1HR INACTIVE(FREE) [0x00000000ffe00000, 0x00000000ffe00000, 0x00000000fff00000]
[0.050s][trace][gc,region   ] G1HR INACTIVE(FREE) [0x00000000fff00000, 0x00000000fff00000, 0x0000000100000000]
[0.050s][debug][gc,ergo,heap] Attempt heap shrinking (CDS archive regions). Total size: 2097152B
[0.050s][trace][gc,region   ] G1HR UNCOMMIT(FREE) [0x00000000ffe00000, 0x00000000ffe00000, 0x00000000fff00000]
[0.050s][trace][gc,region   ] G1HR UNCOMMIT(FREE) [0x00000000fff00000, 0x00000000fff00000, 0x0000000100000000]
[0.057s][trace][gc,region   ] G1HR ALLOC(EDEN) [0x00000000ffd00000, 0x00000000ffd00000, 0x00000000ffe00000]
[0.129s][trace][gc,region   ] G1HR ALLOC(EDEN) [0x00000000ffc00000, 0x00000000ffc00000, 0x00000000ffd00000]
openjdk version "21-internal" 2023-09-19
```

(The `GCCardSizeInBytes` option is there to decrease the minimum heap alignment to 512kb/1M so that setting `-Xms` to an odd value in a later test works)

I.e. the CDS regions are unconditionally uncommitted even through `-Xms == -Xmx`.

The next case just illustrates current (pre-existing) behavior with `-Xms != -Xmx`, showing that CDS regions are always committed, leading to higher than `-Xms` memory usage. I will file an enhancement here, as it is acceptable behavior (to me).

```
$ java -XX:GCCardSizeInBytes=128 -Xmx128m -Xms126m -Xlog:gc+region=trace,gc+ergo+heap=debug -XX:+UseNewCode -version

[0.048s][trace][gc,region   ] G1HR COMMIT(FREE) [0x00000000fff00000, 0x00000000fff00000, 0x0000000100000000]
[0.048s][trace][gc,region   ] G1HR ACTIVE(FREE) [0x00000000fff00000, 0x00000000fff00000, 0x0000000100000000]
[0.048s][debug][gc,ergo,heap] Allocate CDS archive regions. Allocated 2 Committed 2
[0.048s][trace][gc,region   ] G1HR ALLOC(OLD) [0x00000000ffe00000, 0x00000000fff00000, 0x00000000fff00000]
[0.048s][trace][gc,region   ] G1HR ALLOC(OLD) [0x00000000fff00000, 0x00000000fff05f50, 0x0000000100000000]
[0.063s][trace][gc,region   ] G1HR ALLOC(EDEN) [0x00000000ffd00000, 0x00000000ffd00000, 0x00000000ffe00000]
```

I.e. CDS allocation just commits the last few regions without caring about other existing mappings to not fragment the heap.

The next case is the new behavior:
```
java -XX:GCCardSizeInBytes=128 -Xmx128m -Xms128m -Xlog:gc+region=trace,gc+ergo+heap=debug -XX:+UseNewCode -version

[0.047s][trace][gc,region   ] G1HR ACTIVE(FREE) [0x00000000ffe00000, 0x00000000ffe00000, 0x00000000fff00000]
[0.047s][trace][gc,region   ] G1HR ACTIVE(FREE) [0x00000000fff00000, 0x00000000fff00000, 0x0000000100000000]
[0.048s][debug][gc,ergo,heap] Allocate CDS archive regions. Allocated 2 Committed 0
[0.048s][trace][gc,region   ] G1HR ALLOC(OLD) [0x00000000ffe00000, 0x00000000fff00000, 0x00000000fff00000]
[0.048s][trace][gc,region   ] G1HR ALLOC(OLD) [0x00000000fff00000, 0x00000000fff05f50, 0x0000000100000000]
[0.049s][trace][gc,region   ] G1HR CLEANUP(FREE) [0x00000000ffe00000, 0x00000000ffe00000, 0x00000000fff00000]
[0.050s][trace][gc,region   ] G1HR CLEANUP(FREE) [0x00000000fff00000, 0x00000000fff00000, 0x0000000100000000]
[0.050s][debug][gc,ergo,heap] Deallocate CDS archive regions. Freed 2 Uncommitted 0 regions.
[0.057s][trace][gc,region   ] G1HR ALLOC(EDEN) [0x00000000fff00000, 0x00000000fff00000, 0x0000000100000000]
[0.131s][trace][gc,region   ] G1HR ALLOC(EDEN) [0x00000000ffe00000, 0x00000000ffe00000, 0x00000000fff00000]
openjdk version "21-internal" 2023-09-19
```
G1 does not uncommit the region blindly any more.

The next test case shows that if `-Xms` is lower than `-Xmx`, the code still properly uncommits (and no only frees the region).
```
java -XX:GCCardSizeInBytes=128 -Xmx128m -Xms126m -Xlog:gc+region=trace,gc+ergo+heap=debug -XX:+UseNewCode -version

[0.048s][trace][gc,region   ] G1HR COMMIT(FREE) [0x00000000ffe00000, 0x00000000ffe00000, 0x00000000fff00000]
[0.048s][trace][gc,region   ] G1HR ACTIVE(FREE) [0x00000000ffe00000, 0x00000000ffe00000, 0x00000000fff00000]
[0.048s][trace][gc,region   ] G1HR COMMIT(FREE) [0x00000000fff00000, 0x00000000fff00000, 0x0000000100000000]
[0.048s][trace][gc,region   ] G1HR ACTIVE(FREE) [0x00000000fff00000, 0x00000000fff00000, 0x0000000100000000]
[0.048s][debug][gc,ergo,heap] Allocate CDS archive regions. Allocated 2 Committed 2
[0.048s][trace][gc,region   ] G1HR ALLOC(OLD) [0x00000000ffe00000, 0x00000000fff00000, 0x00000000fff00000]
[0.048s][trace][gc,region   ] G1HR ALLOC(OLD) [0x00000000fff00000, 0x00000000fff05f50, 0x0000000100000000]
[0.048s][trace][gc,region   ] G1HR INACTIVE(FREE) [0x00000000ffe00000, 0x00000000ffe00000, 0x00000000fff00000]
[0.049s][trace][gc,region   ] G1HR INACTIVE(FREE) [0x00000000fff00000, 0x00000000fff00000, 0x0000000100000000]
[0.049s][debug][gc,ergo,heap] Deallocate CDS archive regions. Freed 2 Uncommitted 2 regions.
[0.049s][trace][gc,region   ] G1HR UNCOMMIT(FREE) [0x00000000ffe00000, 0x00000000ffe00000, 0x00000000fff00000]
[0.049s][trace][gc,region   ] G1HR UNCOMMIT(FREE) [0x00000000fff00000, 0x00000000fff00000, 0x0000000100000000]
[0.056s][trace][gc,region   ] G1HR ALLOC(EDEN) [0x00000000ffd00000, 0x00000000ffd00000, 0x00000000ffe00000]
[0.124s][trace][gc,region   ] G1HR ALLOC(EDEN) [0x00000000ffc00000, 0x00000000ffc00000, 0x00000000ffd00000]
openjdk version "21-internal" 2023-09-19
```

And finally a "mixed" case where to allocate the CDS archive G1 uses a mix of already allocated regions and committed regions, and uncommits/gives back regions as expected.

```
java -XX:GCCardSizeInBytes=128 -Xmx128m -Xms127m -Xlog:gc+region=trace,gc+ergo+heap=debug -XX:+UseNewCode -version

[0.053s][trace][gc,region   ] G1HR COMMIT(FREE) [0x00000000fff00000, 0x00000000fff00000, 0x0000000100000000]
[0.053s][trace][gc,region   ] G1HR ACTIVE(FREE) [0x00000000fff00000, 0x00000000fff00000, 0x0000000100000000]
[0.053s][debug][gc,ergo,heap] Allocate CDS archive regions. Allocated 2 Committed 1
[0.053s][trace][gc,region   ] G1HR ALLOC(OLD) [0x00000000ffe00000, 0x00000000fff00000, 0x00000000fff00000]
[0.053s][trace][gc,region   ] G1HR ALLOC(OLD) [0x00000000fff00000, 0x00000000fff05f50, 0x0000000100000000]
[0.053s][trace][gc,region   ] G1HR INACTIVE(FREE) [0x00000000ffe00000, 0x00000000ffe00000, 0x00000000fff00000]
[0.054s][trace][gc,region   ] G1HR CLEANUP(FREE) [0x00000000fff00000, 0x00000000fff00000, 0x0000000100000000]
[0.054s][debug][gc,ergo,heap] Deallocate CDS archive regions. Freed 2 Uncommitted 1 regions.
[0.054s][trace][gc,region   ] G1HR UNCOMMIT(FREE) [0x00000000ffe00000, 0x00000000ffe00000, 0x00000000fff00000]
[0.061s][trace][gc,region   ] G1HR ALLOC(EDEN) [0x00000000fff00000, 0x00000000fff00000, 0x0000000100000000]
[0.138s][trace][gc,region   ] G1HR ALLOC(EDEN) [0x00000000ffd00000, 0x00000000ffd00000, 0x00000000ffe00000]
openjdk version "21-internal" 2023-09-19
```

Thanks,
  Thomas

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8232722](https://bugs.openjdk.org/browse/JDK-8232722): G1 archive region deallocation may shrink the heap below -Xms

### Reviewers
 * [Ivan Walulya](https://openjdk.org/census#iwalulya) (@walulyai - **Reviewer**)
 * [Albert Mingkun Yang](https://openjdk.org/census#ayang) (@albertnetymk - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14145/head:pull/14145` \
`$ git checkout pull/14145`

Update a local copy of the PR: \
`$ git checkout pull/14145` \
`$ git pull https://git.openjdk.org/jdk.git pull/14145/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14145`

View PR using the GUI difftool: \
`$ git pr show -t 14145`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14145.diff">https://git.openjdk.org/jdk/pull/14145.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14145#issuecomment-1562614745)

### CODE REVIEW

File: src/hotspot/share/gc/g1/heapRegionManager.cpp [hunk 1]
```diff
@@ -567,10 +568,12 @@ bool HeapRegionManager::allocate_containing_regions(MemRegion range, size_t* com
     if (!curr_region->is_free()) {
```
Comment (walulyai) [inline src/hotspot/share/gc/g1/heapRegionManager.cpp:568]: What happens with regions committed before this false return?

**Review by walulyai — APPROVED** [review]
Lgtm!

File: src/hotspot/share/gc/g1/heapRegionManager.cpp (hunk 1 above)
Comment (tschatzl) [reply src/hotspot/share/gc/g1/heapRegionManager.cpp:568]: I filed https://bugs.openjdk.org/browse/JDK-8309117.

**Comment by ashu-mehra** [comment]
@tschatzl what if the regions allocated for archive objects are not uncommitted but just added to the free list in  `dealloc_archive_regions`? That would also prevent heap going below Xms and would avoid tracking committed regions.
Do you see any issue if the regions are not uncommitted?

File: src/hotspot/share/gc/g1/g1CollectedHeap.cpp [hunk 2]
```diff
@@ -547,23 +547,23 @@ bool G1CollectedHeap::alloc_archive_regions(MemRegion range) {
...
 
   guarantee(reserved.contains(start_address) && reserved.contains(last_address),
             "MemRegion outside of heap [" PTR_FORMAT ", " PTR_FORMAT "]",
             p2i(start_address), p2i(last_address));
 
   // Perform the actual region allocation, exiting if it fails.
   // Then note how much new space we have allocated.
-  if (!_hrm.allocate_containing_regions(range, &commits, workers())) {
+  regions_committed = 0;
+  uint regions_allocated = 0;
```
Comment (albertnetymk) [inline src/hotspot/share/gc/g1/g1CollectedHeap.cpp:558]: Why is alloc-count required? Isn't it sth like `ceil(range/region_size)`?

Comment (tschatzl) [reply src/hotspot/share/gc/g1/g1CollectedHeap.cpp:558]: This adds implicit assumptions about how the range looks like, i.e. the start address; I think it is nicer if the one allocating the memory returns this value like any other information about the allocation.

**Comment by tschatzl** [comment]
> @tschatzl what if the regions allocated for archive objects are not uncommitted but just added to the free list in `dealloc_archive_regions`? That would also prevent heap going below Xms and would avoid tracking committed regions. Do you see any issue if the regions are not uncommitted?

No real issue, but no real advantage either. In case of -Xms != -Xmx this would mean that the `main` method will not start with heap size of -Xms, but additional regions.
Given that CDS is enabled by default, this would effectively mean that unless disabled, the VM would always (effectively) start (at `main`) with an initial heap size a bit larger than requested.

This is allowed (initial heap size is somewhat of a hint), but unexpected I assume.

The question is, should G1 try to honor requested initial heap size or not? (Similar to https://bugs.openjdk.org/browse/JDK-8308854)

File: src/hotspot/share/gc/g1/heapRegionManager.cpp (hunk 1 above)
Comment (tschatzl) [reply src/hotspot/share/gc/g1/heapRegionManager.cpp:568]: Actually I checked/tested again, and the result is that these unnecessarily committed regions will stay around as free regions. No other error or something. I adapted the CR.

File: src/hotspot/share/gc/g1/g1CollectedHeap.cpp (hunk 2 above)
Comment (albertnetymk) [reply src/hotspot/share/gc/g1/g1CollectedHeap.cpp:558]: I think the closure below, `set_region_to_old`, already assumes the start-addr is region-start. In fact, I believe that is an implicit assumption shared by `alloc_archive_regions`, `dealloc_archive_regions`, and `populate_archive_regions_bot_part`.

I'd prefer making it explicit, but YMMV.

**Review by albertnetymk — APPROVED** [review]

**Comment by ashu-mehra** [comment]
> No real issue, but no real advantage either. In case of -Xms != -Xmx this would mean that the main method will not start with heap size of -Xms, but additional regions.

Right, that's the current behavior as well and not uncommitting the regions in `dealloc_archive_regions` doesn't change that. The difference appears only when the archive region mapping fails. If the regions are uncommitted it would bring back heap size to Xms. If not, heap size will remain slightly above Xms, which is same as the current behavior when mapping succeeds.

The advantage is avoiding the need to track committed regions which avoids leaks more G1 specific details to CDS code. Adding GC policy specific details would make it harder to implement https://bugs.openjdk.org/browse/JDK-8296263. I am trying to move G1 specific code out of CDS in https://github.com/openjdk/jdk/pull/14208.
So if not uncommitting simplifies the things then why not?

**Comment by tschatzl** [comment]
> > No real issue, but no real advantage either. In case of -Xms != -Xmx this would mean that the main method will not start with heap size of -Xms, but additional regions.
> 
> Right, that's the current behavior as well and not uncommitting the regions in `dealloc_archive_regions` doesn't change that. The difference appears only when the archive region mapping fails. If the regions are uncommitted it would bring back heap size to Xms. If not, heap size will remain slightly above Xms, which is same as the current behavior when mapping succeeds.

I do not think the current behavior is intended or good. It is at most tolerable. At least it has not been specified as correct in any documentation I could find, and it seems to counter to how initial heap size behaves with other collectors afaik (they just load the data into wherever it lands in old gen and fixes up references).

Hence filing https://bugs.openjdk.org/browse/JDK-8308854.

> The advantage is avoiding the need to track committed regions which avoids leaks more G1 specific details to CDS code. Adding GC policy specific details would make it harder to implement https://bugs.openjdk.org/browse/JDK-8296263. I am trying to move G1 specific code out of CDS in #14208. So if not uncommitting simplifies the things then why not?

That is a strange argument given that it causes some (minor) observable difference in behavior between collectors too. Having collectors behave the same helps reasoning about it.

I will look at the PR later today, haven't noticed it before.

**Comment by tschatzl** [comment]
Maybe it is better to make this change differently, and combine with https://bugs.openjdk.org/browse/JDK-8308854: archive allocation removes/uncommits regions from the heap first, then allocates the new memory. This keeps -Xms. Later deallocation simply frees that memory.

What do you think?

**Comment by ashu-mehra** [comment]
> Maybe it is better to make this change differently, and combine with https://bugs.openjdk.org/browse/JDK-8308854: archive allocation removes/uncommits regions from the heap first, then allocates the new memory. This keeps -Xms. Later deallocation simply frees that memory.

+1 for this approach.

**Comment by tschatzl** [comment]
> > Maybe it is better to make this change differently, and combine with https://bugs.openjdk.org/browse/JDK-8308854: archive allocation removes/uncommits regions from the heap first, then allocates the new memory. This keeps -Xms. Later deallocation simply frees that memory.
> 
> +1 for this approach.

Retracting this change to try the other one.

**Pull request closed without merging**